### DISCUSSION TITLE
color picking??

### DISCUSSION THREAD

**Discussion by istvan-b01** [spakin/SimpInkScr#95] [General]
is there a way to detect the color at a certain coordinate?
I have a png image with very crisp black and white colors which I'd like to scan... I know reading is very different from creating, but if I don't ask I won't know.

**Comment by istvan-b01** [comment]
well, turning it into a bitmap, then checking with a hexeditor to see the structure, extracting the image only, paste it into vim, remove the unneeded, paste it into some language... process... is a way to do it I guess....
any other? :)

**Reply by spakin** [reply]
I'd use [`urllib.request`](https://docs.python.org/3/library/urllib.request.html) to read the image data into a file object, [Pillow](https://pillow.readthedocs.io/en/stable/index.html) to parse the image data, and [NumPy](https://numpy.org/) to speed up processing by converting the image data to an array.  Here's a sample script to get you started:

```Python
import urllib.request
from PIL import Image
import numpy as np

pixel2fill = {False: 'black', True: 'white'}
for obj in all_shapes():
    if obj.tag != 'image':
        continue
    vector_graphic = group()
    href = obj.svg_get('xlink:href', as_str=True)
    fobj = urllib.request.urlopen(href)
    with Image.open(fobj) as im:
        bw_img = im.convert('1')  # Convert to black and white.
        bit_data = np.asarray(bw_img)
        wd, ht = bw_img.width, bw_img.height
        for y in range(ht):
            for x in range(wd):
                bit = bit_data[y, x]
                r = rect((x, y), (x + 1, y + 1), fill=pixel2fill[bit], stroke='none')
                vector_graphic.append(r)
```

The above is designed for embedded images.  It likely requires some modifications for external images.

**Reply by istvan-b01** [reply]
Wow, thank you very much, this is a good source to learn from as well.
Regarding the color picking, since the images were pngs, therefore they did scale up nicely, I loaded them into gimp, minimized the number of colors to 2 and used brush to fill to a certain one that I would recogize in hex like (rgb(12,34, 56). Then opened the saved image in a hex editor that allowed me to retrieve the data in text format and then in vim I eliminated the data that was before the first occurrence of the 12, 34, 56, scaled  it down so one rgb value represented only 1 pixel in the resolution, then turned all that into zeros and ones. After this breaking down this long stream of data to the width and height of the image, I did see through the ones and zeros the information I was seeking. Fun fact, BMP images are stored upside down for some reason, so after the image data the last line appears then the one above and so one. vim's macros being a miracle, I flipped that and there I had the information I needed in binary. breaking that further into character size (DOS characters  being 16x9 (nine!!!!) is against the fabric of the universe I believe!!), than I just used python to turn them into hexes so I could feed it to the little script I included to generate them in inkscape. So as you see I just put on some elbow grease and did it in the hard way, but technically it works as you can turn any image into a BMP and process the data in it. Nonetheless, I most certainly want to have to nice script that can do all this, so the script you included is very helpful to understand the underlying structure of inkscape and also how pythong works. Thank you!

I don't know if you noticed that I replied in the other post stating that i don't have much luck with the 'save_file(name)' feature as it didn't work for me. it's only a convenience but as I mentioned in the other post, is it possible to include a path/directory to the filename in order to save multiple files to different places. I did try and it gave the file not found error despite the the path I gave was correct. But when I removed that and just simply wanted to save the current layer into a file, then reusing the layer again, it did not work, although that time there was no error message either. I tried to delete (del layername) the layer after each usage to keep it as a single layer for each file, but none of that worked. all the shapes crowded up till the end (although after deletion they were no longer referable)  and no file was saved what so ever.
what was I doing wrong if you would please point it out.

**Reply by istvan-b01** [reply]
I closed the other discussion as I found the answer for deleting objects as well and I wasn't aware that the discussion will disappear.
Anyway, I finished it with a big thank you for assisting me with my project. I learnt a great deal and made me being able to carry on with my project of which the retro fonts I tried to recreate was just a small part. Anyway, yours is inspiring and wish great success with its further development!
